src/algebra/order/hom/monoid.lean

```diff
@@ -0,0 +1,481 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import algebra.group.hom
+import algebra.order.with_zero
+import order.hom.basic
+
+/-!
+# Ordered monoid and group homomorphisms
+
+This file defines morphisms between (additive) ordered monoids.
+
+## Types of morphisms
+
+* `order_add_monoid_hom`: Ordered additive monoid homomorphisms.
+* `order_monoid_hom`: Ordered monoid homomorphisms.
+* `order_monoid_with_zero_hom`: Ordered monoid with zero homomorphisms.
+
+## Typeclasses
+
+* `order_add_monoid_hom_class`
+* `order_monoid_hom_class`
+* `order_monoid_with_zero_hom_class`
+
+## Notation
+
+* `→+o`: Bundled ordered additive monoid homs. Also use for additive groups homs.
+* `→*o`: Bundled ordered monoid homs. Also use for groups homs.
+* `→*₀o`: Bundled ordered monoid with zero homs. Also use for groups with zero homs.
+
+## Implementation notes
+
+There's a coercion from bundled homs to fun, and the canonical notation is to use the bundled hom as
+a function via this coercion.
+
+There is no `order_group_hom` -- the idea is that `order_monoid_hom` is used.
+The constructor for `order_monoid_hom` needs a proof of `map_one` as well as `map_mul`; a separate
+constructor `order_monoid_hom.mk'` will construct ordered group homs (i.e. ordered monoid homs
+between ordered groups) given only a proof that multiplication is preserved,
+
+Implicit `{}` brackets are often used instead of type class `[]` brackets. This is done when the
+instances can be inferred because they are implicit arguments to the type `order_monoid_hom`. When
+they can be inferred from the type it is faster to use this method than to use type class inference.
+
+## Tags
+
+ordered monoid, ordered group, monoid with zero
+-/
+
+open function
+
+variables {F α β γ δ : Type*}
+
+infixr ` →*₀ `:25 := monoid_with_zero_hom
+
+instance [preorder α] [preorder β] [order_hom_class F α β] : has_coe_t F (α →o β) :=
+⟨λ f, { to_fun := f, monotone' := order_hom_class.mono _ }⟩
```

Comment:
I think this is a good idea and wanted to apply it some point in the future for other morphisms. Could you move it to somewhere closer to the definition of `order_hom`?